src/tapedfunction.jl

```diff
@@ -168,127 +156,184 @@ function (instr::Instruction{F})() where F
-arg_boxer(var::IRTools.Variable, boxes::Dict{IRTools.Variable, Box{Any}}) =
-    get!(boxes, var, Box{Any}(nothing))
-function args_initializer(ins::BlockInstruction)
+function args_initializer(taped::Taped, all_boxes::Dict{Symbol, Box{<:Any}})
+    funcbox = _find_slot(all_boxes, 1)
+    argsbox = [_find_slot(all_boxes, i + 1) for i in 1:taped.arity]
     return (args...) -> begin
-        @assert length(args) + 1 == length(ins.args)
-        ins.args[1].val = ins.tape.func
+        funcbox.val = taped.func
+        @assert length(args) == taped.arity
         for i in 1:length(args)
-            ins.args[i + 1].val = val(args[i]) # fill the boxes
+            argsbox[i].val = val(args[i]) # fill the boxes
         end
     end
 end
 
-function translate!(taped::Taped, ir::IRTools.IR)
+function translate!(taped::Taped, ir::Core.CodeInfo)
     tape = taped.tape
-    boxes = Dict{IRTools.Variable, Box{Any}}()
+    boxes = Dict{Symbol, Box{<:Any}}()
     _box = (x) -> arg_boxer(x, boxes)
-    for (blk_id, blk) in enumerate(IRTools.blocks(ir))
-        # blocks
-        blk_args = IRTools.arguments(blk)
-        push!(tape, BlockInstruction(blk_id, map(_box, blk_args), taped))
-        # `+ 1` because we will have an extra ins at the beginning
-        taped.block_map[blk_id] = length(tape) + 1
-        # normal instructions
-        for (x, st) in blk
-            if Meta.isexpr(st.expr, :call)
-                args = map(_box, st.expr.args)
-                # args[1] is the function
-                f = args[1]
-                ins = Instruction(f, args[2:end] |> Tuple,
-                                  _box(x), taped)
-                push!(tape, ins)
-            elseif Meta.isexpr(st.expr, :new)
-                args = map(_box, st.expr.args)
-                ins = Instruction(__new__, args |> Tuple, _box(x), taped)
-                push!(tape, ins)
-            elseif isa(st.expr, GlobalRef)
-                ins = Instruction(val, (st.expr,), _box(x), taped)
-                push!(tape, ins)
-            else
-                @error "Unknown IR code: " st
-            end
-        end
 
-        # branches (including `return`)
-        for br in IRTools.branches(blk)
-            if br.condition === nothing && br.block == 0 # a return
-                ins = ReturnInstruction(_box(br.args[1]), taped)
-                push!(tape, ins)
-            else
-                cond = br.condition === nothing ? false : _box(br.condition)
-                isa(cond, Bool) && (cond = Box{Any}(cond)) # unify the condiftion type
-                ins = BranchInstruction(
-                    cond, br.block, map(_box, br.args), taped)
-                push!(tape, ins)
-            end
-        end
+    builder = InstructionBuilder(taped, ir, _box)
+
+    for (idx, line) in enumerate(ir.code)
+        isa(line, Core.Const) && (line = line.val) # unbox Core.Const
+        builder(Core.SSAValue(idx), line)
```

Comment:
I think a simpler design would be
```suggestion
    for (idx, line) in enumerate(ir.code)
        push!(taped, _instruction(Core.SSAValue(idx), _unbox(line), taped, ir, boxes))
```
since then you could remove the `InstructionBuilder`, avoid the unpacking in every function, and could remove the `push!` in all functions.

Additionally, one would define
```julia
_unbox(line) = line
_unbox(line::Core.Const) = line.val
```